examples/cms/src/App.tsx

```diff
@@ -0,0 +1,86 @@
+import "./App.css"
+
+import { framer, ManagedCollection } from "framer-plugin"
+import { useEffect, useLayoutEffect, useState, useRef } from "react"
+import { DataSource, getDataSource, getDataSourcesIds } from "./data"
+import { FieldMapping } from "./FieldMapping"
+import { SelectDataSource } from "./SelectDataSource"
+import { Spinner } from "./components/Spinner"
+
+interface AppProps {
+    collection: ManagedCollection
+    previouslyConfiguredDataSourceId: string | null
+    previouslyConfiguredSlugFieldId: string | null
+}
+
+export function App({ collection, previouslyConfiguredDataSourceId, previouslyConfiguredSlugFieldId }: AppProps) {
+    const [dataSource, setDataSource] = useState<DataSource | null>(null)
+
+    const hasStartedLoadingDataSource = useRef(false)
+    const [isLoadingConfiguredDataSource, setIsLoadingConfiguredDataSource] = useState(
+        Boolean(previouslyConfiguredDataSourceId)
+    )
+    const hasDataSourceSelected = Boolean(isLoadingConfiguredDataSource || dataSource)
+
+    useLayoutEffect(() => {
+        framer.showUI({
+            width: hasDataSourceSelected ? 360 : 320,
+            height: hasDataSourceSelected ? 425 : 305,
+            minWidth: hasDataSourceSelected ? 360 : undefined,
+            minHeight: hasDataSourceSelected ? 425 : undefined,
+            resizable: dataSource !== null,
+        })
+    }, [hasDataSourceSelected, dataSource])
+
+    useEffect(() => {
+        if (!previouslyConfiguredDataSourceId || hasStartedLoadingDataSource.current) {
+            return
+        }
```

Comment:
Why do you need the logic for `hasStartedLoadingDataSource`? The value of `previouslyConfiguredDataSourceId` can't change at runtime, right?

Reply:
The ref is needed because in React's strict mode (development), useEffect runs twice by design. Without `hasStartedLoadingDataSource`, you'd fetch the data twice. While `previouslyConfiguredDataSourceId` won't change, the ref prevents this double-fetch.

Reply:
React is running the effect twice, to make sure you clean up your effects properly. They definitely don't want you to add workaround like these

Reply:
You can cancel the first fetch using an abort controller